torch/csrc/autograd/variable.h

```diff
@@ -1,329 +1,615 @@
+  /// Variable.
+  void rebase_history(Edge gradient_edge);
 
-  // The base Variable (never a view)
+  /// The base `Variable` (never a view).
   Variable base;
 
-  // The value of the version_counter at the time grad_fn was created. The
-  // _grad_fn field is stale if attr_version != version_counter.current_version()
-  int attr_version;
+  /// The value of the version_counter at the time grad_fn was created. The
+  /// grad_fn field is stale if attr_version !=
+  /// version_counter.current_version().
+  uint32_t attr_version;
 };
 
-inline Variable make_variable(at::Tensor data, bool requires_grad=false) {
-  if (!data.defined()) {
-    return Variable();
-  }
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//                        Variable Implementation
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+namespace detail {
+inline at::Tensor handle_scalars(at::Tensor& data) {
 #ifndef WITH_SCALARS
   if (data.dim() == 0) {
-    // don't expose 0-dim tensors to Variable API.
-    data = data.as_strided_({1}, {1});
+    // Don't expose 0-dim tensors to Variable API.
+    return data.as_strided_({1}, {1});
   }
 #endif
+  return data;
+}
+} // namespace detail
+
+// Factory Functions
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+inline Variable make_variable_view(
+    Variable base,
+    at::Tensor data,
+    Edge gradient_edge = Edge()) {
+  if (data.defined()) {
+    data = detail::handle_scalars(data);
+    auto impl = new Variable::ViewImpl(
+        std::move(base), std::move(data), std::move(gradient_edge));
+    return Variable(impl, /*retain=*/false);
+  }
+  return Variable();
+}
 
-  return Variable(new VariableImpl(std::move(data), requires_grad), false);
+inline Variable make_variable(at::Tensor data, bool requires_grad) {
+  if (data.defined()) {
+    auto impl = new Variable::Impl(detail::handle_scalars(data), requires_grad);
+    return Variable(impl, /*retain=*/false);
+  }
+  return Variable();
 }
 
-inline Variable make_variable(at::Tensor data, int output_nr, std::shared_ptr<Function> grad_fn) {
-  if (!data.defined()) {
-    return Variable();
+inline Variable make_variable(at::Tensor data, Edge gradient_edge) {
+  if (data.defined()) {
+    auto impl = new Variable::Impl(
+        detail::handle_scalars(data), false, std::move(gradient_edge));
+    return Variable(impl, /*retain=*/false);
   }
+  return Variable();
+}
 
-#ifndef WITH_SCALARS
-  if (data.dim() == 0) {
-    // don't expose 0-dim tensors to Variable API.
-    data = data.as_strided_({1}, {1});
+// Tensor Conversion
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+inline Variable& as_variable_ref(at::Tensor& tensor) {
+#ifdef DEBUG
+  // dynamic_cast will return a nullptr if the `TensorImpl`'s dynamic type is
+  // not `Variable::Impl`.
+  if (dynamic_cast<Variable::Impl*>(tensor.get()) == nullptr) {
```

Comment:
This can be a `ViewImpl` too right?

Reply:
Should be fine I believe because `ViewImpl` inherits from `Impl` so `dynamic_cast`ing a `ViewImpl` to `Impl` would still work (i.e. return non-`nullptr`)

Reply:
Ah I wasn't aware of the inheritance. Never mind the comment then